src/httpserver.cpp

```diff
@@ -650,6 +658,97 @@ HTTPRequest::RequestMethod HTTPRequest::GetRequestMethod()
     }
 }
 
+bool HTTPRequest::StartStreaming(int nStatus)
+{
+    assert(!streaming);
+    streaming = new StreamingData();
+    HTTPEvent* ev = new HTTPEvent(eventBase, true, boost::bind(evhttp_send_reply_start, req, nStatus, (const char*)NULL));
+    ev->trigger(0);
+    return true;
+}
+
+/** TODO: move HTTPRequest::StreamingData to another unit */
+#define MAX_CHUNK_BUFFER (256*1024)
+
+HTTPRequest::StreamingData::StreamingData():
+    buffer_bytes(0)
+{
+    databuf = evbuffer_new();
+    assert(databuf);
+}
+
+HTTPRequest::StreamingData::~StreamingData()
+{
+    evbuffer_free(databuf);
+}
+
+/** Update buffer length in synchronization structure.
+ * @note: Can only safely read the length of the buffer from the event thread.
+ */
+void HTTPRequest::StreamingData::Update(struct evhttp_connection* evcon)
+{
+    {
+        std::unique_lock<std::mutex> lock(cs);
+        struct bufferevent* bufev = evhttp_connection_get_bufferevent(evcon);
+        assert(bufev);
+        struct evbuffer *output = bufferevent_get_output(bufev);
+        assert(output);
+        buffer_bytes = evbuffer_get_length(output);
+    }
+    cond.notify_all();
+}
+
+/** Called by evhttp from event thread after a chunk has been written.
+ */
+void HTTPRequest::StreamingData::http_chunk_cb(struct evhttp_connection* evcon, void* arg)
+{
+    HTTPRequest::StreamingData *self = (HTTPRequest::StreamingData*) arg;
+    self->Update(evcon);
+    // LogPrintf("http_chunk_cb %i\n", self->buffer_bytes);
+}
+
+/** Send current chunk in databuf.
+ * @note: Can only safely be called from event thread
+ */
+void HTTPRequest::StreamingData::SendChunk(struct evhttp_request* req)
+{
+    // LogPrintf("set_http_chunk_cb\n");
+    {
+        std::unique_lock<std::mutex> lock(cs);
+        evhttp_send_reply_chunk_with_cb(req, databuf, &http_chunk_cb, this);
```

Comment:
Yes, I need to find exactly what version this was introduce in and guard the streaming stuff with `#if LIBEVENT_VERSION_NUMBER >= 0x0201XXXX`. I think it can only be supported for newer libevent .